### PR TITLE
CI/CD: Add a GitHub Action Bot to tag PRs with merge conflicts 

### LINKED ISSUE #160: CI/CD: Add a GitHub Action Bot to tag PRs with merge conflicts
## Description

Add a GitHub Action Bot to tag PRs with merge conflicts. The Bot should do the following: 

- Tag a PR with `merge-conflicts` label if it runs into a merge conflict. Remove it up once it has been resolved.
- Comment instructions to resolve merge conflicts by quoting GitHub Docs. 
- Remind the PR author every 24 hours if the merge conflicts is not fixed.

### PR DESCRIPTION
**Pull request by rubyruins**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

I have created a workflow that runs the following 2 jobs sequentially:
- Checks for merge conflicts, and if present, labels the PR with the `merge-conflicts` label
- Checks if the `merge-conflicts` label is present, and if so, comments with appropriate guidelines to resolve it
- I have not been able to implement the last functionality yet:
> Remind the PR author every 24 hours if the merge conflict is not fixed

As discussed in #160, I'm willing to discuss how to approach and create this feature. Any resources would be really helpful.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.

## Related Issues or Pull Requests

Fixes #160

### CODE REVIEW

**Comment by rubyruins** [comment]
Also, currently the Check for merge conflicts workflow is failing since it requires a `merge-conflicts` label to be created in the repository before the action can be used.

**Comment by HarshCasper** [comment]
Seems like I broke your CI again :sweat_smile: 

I updated the action to trigger the build again. And it seems it's still not working. 

Adding @vybhav72954 here who can help you debug this.

**Comment by rubyruins** [comment]
Sure, thanks. New error this time, resource not accessible 😅

**Comment by vybhav72954** [comment]
@rubyruins Thank you for coming up with a good implementation. Have you tested this action remotely on any of your Repo?

**Comment by rubyruins** [comment]
Thanks, @vybhav72954! Yep, I have tried this on my end and have noticed these things -

- In case of a merge conflict on making a PR, workflows do not get triggered. The thing is, I made 2 branches on a private repo, created a conflict between them, and tested a PR from the second branch to the first. In this case, the workflow should get triggered and add the label. 
- What is happening in my case is that since I am the repo's author as well as the person making the PR, it gives me the Github message saying that merging is blocked, instead of letting the action run and label it on its own. I don't know if this is because I own the repo vs the expected scenario that I create a PR on a repo I do not own. 
- According to Google search, Github actions do not seem to get triggered on any kind of merge conflicts, but some people say it does get triggered for some cases. So I'm not sure about it here. For the action I tried, I saw many example screenshots that say it does get triggered even in case of merge conflicts, so I am hoping that is the case here. However, since the repo I tried on was owned by me, it could be an exception. 
- I tried checking if the second job works, assuming that the first job works on a repo not owned by me when I make a PR with merge conflicts. I manually created a PR and added the merge-conflicts label to it, since the second job which gets triggered after the first one compares the labels to see if the merge-conflicts label is there. I can definitely confirm this works, and have added a screenshot below. 
image
- In the above case, there are 2 things to note-
- a. The PR I made in this case had no conflicts. I manually added it on my own since it did not seem to get added on its own (see above).
- b. At this point, the jobs were not sequential. So the label checker completed first, found the relevant label, and commented. The first job completed after that (which checks for conflicts), and I can say that it worked in a way, since as there were no conflicting changes, it removed them on its own, which is the expected behavior of that action.

So yes, I'm a bit unsure too if this works! I have noticed it partly working at least, for my own repo. I'm also relatively new to Github actions, so I had to read up a bit to figure out how to combine some existing actions for our use case. Thanks a lot for taking the time to review :) I'm open to any guidance that can help us figure this out!

**Comment by rubyruins** [comment]
> The thing is GraphQL API is a bit funny, rather than as a credential, action tuns as a user (this might not be the case here, this is just a preliminary guess). As this action is currently running from the fork created for this PR, it will fail because the OP doesn't have permission to access this repo (Conclusion is based on the `resource not accessible` error).

Yes, I think this is possible too. As a fork, this workflow may not be able to access the repo's secrets, so the action may be failing.

**Comment by vybhav72954** [comment]
> Thanks, @vybhav72954! Yep, I have tried this on my end and have noticed these things -
> 
> - In case of a merge conflict on making a PR, workflows do not get triggered. The thing is, I made 2 branches on a private repo, created a conflict between them, and tested a PR from the second branch to the first. In this case, the workflow should get triggered and add the label. 
> - What is happening in my case is that since I am the repo's author as well as the person making the PR, it gives me the Github message saying that merging is blocked, instead of letting the action run and label it on its own. I don't know if this is because I own the repo vs the expected scenario that I create a PR on a repo I do not own. 
> - According to Google search, Github actions do not seem to get triggered on any kind of merge conflicts, but some people say it does get triggered for some cases. So I'm not sure about it here. For the action I tried, I saw many example screenshots that say it does get triggered even in case of merge conflicts, so I am hoping that is the case here. However, since the repo I tried on was owned by me, it could be an exception. 
> - I tried checking if the second job works, assuming that the first job works on a repo not owned by me when I make a PR with merge conflicts. I manually created a PR and added the merge-conflicts label to it, since the second job which gets triggered after the first one compares the labels to see if the merge-conflicts label is there. I can definitely confirm this works, and have added a screenshot below. 
> image
> - In the above case, there are 2 things to note-
> - a. The PR I made in this case had no conflicts. I manually added it on my own since it did not seem to get added on its own (see above).
> - b. At this point, the jobs were not sequential. So the label checker completed first, found the relevant label, and commented. The first job completed after that (which checks for conflicts), and I can say that it worked in a way, since as there were no conflicting changes, it removed them on its own, which is the expected behavior of that action.
> 
> So yes, I'm a bit unsure too if this works! I have noticed it partly working at least, for my own repo. I'm also relatively new to Github actions, so I had to read up a bit to figure out how to combine some existing actions for our use case. Thanks a lot for taking the time to review :) I'm open to any guidance that can help us figure this out!
> 

Got it, that's some good piece of information for me. The thing is i have a second GitHub account for this purpose only XD.

So i will check if the action is doing what it's supposed to do in the case or not. Some clarification here.

For the merged conflict trigger. GitHub loves `async` jobs and hence there is no way to trigger that, you can nust make a check for it.

Let me play a but with the action from my end. I will come back to you asap!

**Comment by rubyruins** [comment]
@vybhav72954, that sounds good! Great idea :) sure, take your time.

**Comment by rubyruins** [comment]
Hi @vybhav72954, any luck with the action?

**Comment by rubyruins** [comment]
@HarshCasper @vybhav72954, just a reminder :) do let me know if I can help in any way.

**Comment by vybhav72954** [comment]
> @HarshCasper @vybhav72954, just a reminder :) do let me know if I can help in any way.

I am working on this as kf writing this.👍🏻

**Comment by vybhav72954** [comment]
> > @HarshCasper @vybhav72954, just a reminder :) do let me know if I can help in any way.
> 
> I am working on this as kf writing this.👍🏻

I tried this Action on multiple instances.

- On Repo's owned by me.
- Repo not owned by me but I have been invited as a collaborator.
- On forks

My final takeaway is that the error faced would be sorted out upon merging the PR, though there are a few probable vulnerabilities, regarding the Github Token here. I will post my final changes (if any) by tomorrow for sure.

**Comment by rubyruins** [comment]
Sounds good! Thank you for taking the time to test it :)

**Comment by rubyruins** [comment]
@vybhav72954, were you able to fix the remaining issues? Is the PR good to go? 😅

**Comment by vybhav72954** [comment]
@rubyruins I tried working on the parallel jobs thing you mentioned. I was unable to come up with something substantial.
Let's keep it a WIP while we can merge this and have a first-hand look at the working action.

Though a few things I will recommend:

File: .github/workflows/conflicts.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: 'Merge Conflict Watcher'
+on:
+  pull_request:
+    branches:
+      - main
```
Comment (vybhav72954) [inline .github/workflows/conflicts.yml:5]: Add more event triggers, like `synchronize` and `edited` is a must.

Use this [link](https://docs.github.com/en/enterprise-server@2.22/actions/reference/events-that-trigger-workflows#pull_request)

File: .github/workflows/conflicts.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+name: 'Merge Conflict Watcher'
+on:
+  pull_request:
```
Comment (vybhav72954) [inline .github/workflows/conflicts.yml:3]: This trigger will help as well

```
on:
  pull_request_review_comment:
    types: [created, deleted]
```

**Review by vybhav72954 — CHANGES_REQUESTED** [review]
@HarshCasper Rest LGTM

**Comment by rubyruins** [comment]
Hi @vybhav72954, I've made the changes. Does it look good now?

File: .github/workflows/conflicts.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    types: [synchronize, edited]
```
Comment (vybhav72954) [inline .github/workflows/conflicts.yml:4]: Shouldn't have removed `opened`

Final change

**Review by vybhav72954 — CHANGES_REQUESTED** [review]

File: .github/workflows/conflicts.yml (hunk 3 above)
Comment (rubyruins) [reply .github/workflows/conflicts.yml]: > Shouldn't have removed `opened`

Oops, sorry. I've added it now :)

**Review by vybhav72954 — APPROVED** [review]
@HarshCasper LGTM
@rubyruins Thank you for the Contribution. Will ping you if any bug arises in any case XD

**Comment by rubyruins** [comment]
@vybhav72954, haha, sure 😅 thanks a lot for helping me out with this one!

**Comment by rubyruins** [comment]
Also, what should we do about the 'remind 24 hours part?' I have tried looking for resources online, but wasn't able to find anything useful!

**Comment by rubyruins** [comment]
@HarshCasper @vybhav72954

**Comment by rubyruins** [comment]
@HarshCasper @vybhav72954 could you please let me know what must be done next in order to merge the PR?

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**